### PR TITLE
Add an enhancement proposal for the "always" policy automation mode

### PR DESCRIPTION
**Pull request by mprahl**
Relates:
https://github.com/open-cluster-management-io/enhancements/issues/59

### CODE REVIEW

**Comment by mprahl** [comment]
@gparvin could you please review this?

File: enhancements/sig-policy/59-ansible-always-mode/README.md [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+## Motivation
+
+The Open Cluster Management (OCM) Policy addon supports running Ansible jobs on policy violations. The Policy addon
+currently supports the run modes `once` and `disabled`. The `once` mode is limiting since after a policy violation
+occurs, the Ansible job runs once and then it is disabled. A user must manually renable it for the Ansible job to run
+again on the next policy violation.
+
+An additional mode of `always` should be added so that an Ansible job can be run for every unique policy violation per
+managed cluster. A simple example is that a user might use the OCM Policy addon Ansible integration to create tickets in
+their ticketing system on every policy violation per cluster. The new `always` mode would enable this use-case.
```
Comment (JustinKuli) [inline enhancements/sig-policy/59-ansible-always-mode/README.md:26]: I'm wondering if `always` is the right word for this - it's not like this automation will be continuously running under this proposal. Something like `onEveryViolation` might be more accurate.

And it opens the question of: what if a user wants to do something when a policy becomes compliant on a cluster? Eg, if an automation should open a ticket when something is noncompliant, maybe I want another automation to close a ticket when it becomes compliant?

Comment (mprahl) [reply enhancements/sig-policy/59-ansible-always-mode/README.md:26]: We agreed on `everyEvent`.

Running automation on other states is a good idea but out of scope of this enhancement since it's handled by the `eventHook` field on `PolicyAutomation` objects.

File: enhancements/sig-policy/59-ansible-always-mode/README.md [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+```
+
+Several minutes later, the `enable-etcd-encryption` policy becomes compliant on `cluster1`. The Governance Policy
+Propagator controller would not run an Ansible job and set the following status on the `PolicyAutomation` object.
+
+```yaml
+status:
+  ranOnClusters:
+    - cluster2
+```
```
[the remark this answers is not in the document]
Comment (mprahl) [reply enhancements/sig-policy/59-ansible-always-mode/README.md:98]: You can check Ansible Tower/AAP itself or see the `AnsibleJobs` that were created by the Governance Policy Propagator but I believe the latter get cleaned up periodically. I'd say this is out of scope of this enhancement though.

[the remark this answers is not in the document]
Comment (JustinKuli) [reply enhancements/sig-policy/59-ansible-always-mode/README.md]: I just meant, this could be good context for someone looking at the proposal later.

[the remark this answers is not in the document]
Comment (mprahl) [reply enhancements/sig-policy/59-ansible-always-mode/README.md]: Thanks! I added that to the doc.

File: enhancements/sig-policy/59-ansible-always-mode/README.md [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+Now consider the case where the user made a change to their Ansible job and wants the Governance Policy Propagator
+controller to rerun the Ansible jobs for all current noncompliant managed clusters. The user would set the existing
+annotation of `policy.open-cluster-management.io/rerun` to `true` on the `PolicyAutomation` object and thus signal the
+Governance Policy Propagator controller to ignore `status.ranOnClusters` and run an Ansible job for every noncompliant
+managed cluster.
```
[the remark this answers is not in the document]
Comment (mprahl) [reply enhancements/sig-policy/59-ansible-always-mode/README.md]: I added this as an additional user story. Thanks!

File: enhancements/sig-policy/59-ansible-everyevent-mode/README.md [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+This status update would tell the Governance Policy Propagator controller to not run the Ansible job again on `cluster1`
+the next time it notices that the `enable-etcd-encryption` policy is noncompliant on `cluster1` such as during periodic
+reconciling or status updates. This isn't meant to store a complete history of which clusters had an Ansible job run due
+to a policy violation. It is meant to track the clusters **currently** in the state based on the `eventHook` value in
+the `PolicyAutomation` object and the appropriate automation was run for those clusters. Such a history can be found on
+the Ansible Tower/Ansible Automation Platform job history.
+
+Several minutes later, the `enable-etcd-encryption` policy becomes noncompliant on `cluster2`. The Governance Policy
+Propagator controller would run the Ansible job from the Hub while passing the `extra_var` of
+`target_cluster: cluster2`. It would then set the following status on the `PolicyAutomation` object.
```
Comment (gparvin) [inline enhancements/sig-policy/59-ansible-everyevent-mode/README.md:90]: Have you thought about a "mass noncompliance" scenario?  What if something happens (maybe a bad edit to the policy) that causes a currently fully compliant policy to suddenly become NonCompliant across all managed clusters?  If there are a large number of managed clusters, are we creating a large number of automations - 1 for each managed cluster?

Comment (gparvin) [reply enhancements/sig-policy/59-ansible-everyevent-mode/README.md:90]: Maybe a realistic scenario would be using hub templates to push a certificate from the hub to managed clusters.  Now that certificate expires -- causing a resource managed by a policy to become noncompliant everywhere it's deployed, I wonder if you can add some comments to help understand that multiple clusters becoming noncompliant at once are properly handled.

Comment (mprahl) [reply enhancements/sig-policy/59-ansible-everyevent-mode/README.md:90]: @gparvin yes, it'd be one Ansible job per cluster. I don't think it'd overwhelm the Ansible Tower/AAP much more than if we consolidated them, but it could potentially overwhelm the Ansible Operator and create a lot of objects in etcd.

Alternatively, we could continue using `target_clusters: [...]` as the extra var and group them together when we can. I was originally thinking `target_cluster` would be an easier interface to work with, but I'll go back to `target_clusters: [...]` to be safe.

**Comment by mprahl** [comment]
@gparvin and @JustinKuli and @ChunxiAlexLuo could you please take another look with the feedback incorporated from the call earlier?

**Review by JustinKuli — CHANGES_REQUESTED** [review]

File: enhancements/sig-policy/59-ansible-everyevent-mode/README.md [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+annotation of `policy.open-cluster-management.io/rerun` to `true` on the `PolicyAutomation` object and thus signal the
+Governance Policy Propagator controller to ignore `status.clustersWithEvent` and run an Ansible job for every
+noncompliant managed cluster.
+
+### Limiting the Number of Ansible Jobs
+
+If a user has a policy that they are concerned with often going between compliant and non-compliant, they may not want
+to run an Ansible job every time the cluster becomes noncompliant in order to not overwhelm Ansible Tower/AAP. In this
+case, a new optional spec field of `delayAfterRunSeconds` (follows the
+[API naming conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#naming-conventions))
```
[the remark this answers is not in the document]
Comment (JustinKuli) [reply enhancements/sig-policy/59-ansible-everyevent-mode/README.md:125]: `repeatEventDelaySeconds` could be a consideration? I don't know if it's any better in general, but it might clarify the behavior I commented on separately. Eg, this would be more clear that the triggering event `compliant->noncompliant` has to occur after this delay - ***if*** that's even the behavior we want, of course.

File: enhancements/sig-policy/59-ansible-everyevent-mode/README.md [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+tickets in their ticketing system on every policy violation per cluster. The new `everyEvent` mode would enable this
+use-case.
+
+### Goals
+
+1. The Policy addon is able to run an Ansible job on every unique policy violation per managed cluster.
+
+### Non-Goals
+
+1. Consolidating Ansible job runs when multiple managed clusters are in violation.
```
[the remark this answers is not in the document]
Comment (mprahl) [reply enhancements/sig-policy/59-ansible-everyevent-mode/README.md:37]: You're right. I forgot to remove this. The behavior of grouping will remain as it is today.

File: enhancements/sig-policy/59-ansible-everyevent-mode/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (mprahl) [reply enhancements/sig-policy/59-ansible-everyevent-mode/README.md:130]: I clarified the behavior in the doc. Could you please take another look?

File: enhancements/sig-policy/59-ansible-everyevent-mode/README.md [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+[API naming conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#naming-conventions))
+would be added so the user could specify the minimum amount of seconds before an automation can be retriggered on the
+same cluster. This would default to 0 seconds.
+
+If a user sets `delayAfterRunSeconds` to `600` for example, and the cluster becomes noncompliant a second time and stays
+noncompliant past the 600 seconds, the Ansible job would be run again for that cluster. When the `delayAfterRunSeconds`
+value is set, the `PolicyAutomation` controller would not remove the `status.clustersWithEvent` entry for the cluster
+until it has become compliant and the number of seconds set in `delayAfterRunSeconds` have elapsed since the
+`automationStartTime` value. From an implementation point of view, this would mean requeing the reconcile request to
+that point in time.
```
[the remark this answers is not in the document]
Comment (JustinKuli) [reply enhancements/sig-policy/59-ansible-everyevent-mode/README.md]: The addition of `eventTime` should let it work as described, so that is good.

I still think there's the potential here that AutomationJobs will be created on a loop. The frequency of that loop is limited by the delay here, but I think the alternate behavior I described could prevent that loop.

The question is: if the policy returns to noncompliance before the delay is up, should the automation just be run again? I think there is a case to made for "no" - in this case the automation didn't really work, and it might be bad to just re-run it because that creates more objects in k8s.

I'd like to see it clarified in the goals/non-goals of the proposal. As you now describe it, I think the goal is to limit the rate that AnsibleJobs are created, with a non-goal of preventing loops from creating an infinite number of Jobs.

[the remark this answers is not in the document]
Comment (mprahl) [reply enhancements/sig-policy/59-ansible-everyevent-mode/README.md]: @JustinKuli right, this behavior is as intended because the Ansible job likely won't resolve the policy violation but merely just trigger some external action such as creating a service now ticket or creating some kind of alert, which we would want to run repeatedly. So the goal is to just reduce the amount of Ansible jobs that get created for an unstable policy.

I added this in the goal and non-goal section as you suggested.

**Comment by mprahl** [comment]
@JustinKuli @ChunxiAlexLuo @gparvin is this good to merge?

**Review by ChunxiAlexLuo — APPROVED** [review]
/lgtm

/hold for @JustinKuli

File: enhancements/sig-policy/59-ansible-everyevent-mode/README.md [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+1. The Policy addon is able to run an Ansible job on every unique policy violation per managed cluster.
+1. Provide an option to avoid excessive Ansible job creation on an unstable policy (e.g. frequest compliant ->
+   noncompliant -> compliant).
+
+### Non-Goals
+
+1. Keeping track of every historical `PolicyAutomation` run.
+1. Adding additional `eventHook` options.
+1. Avoiding infinite creation of Ansible jobs on unstable policies (e.g. frequest compliant -> noncompliant ->
+   compliant).
```
Comment (JustinKuli) [inline enhancements/sig-policy/59-ansible-everyevent-mode/README.md:42]: @gparvin, I'd like to get your opinion on this.

The situation is: if the PolicyAutomation takes an action on the cluster which resolves the violation, but the policy becomes non-compliant again within the specified `delayAfterRunSeconds`, then the PolicyAutomation will be run again. Potentially in a loop.

As an alternative, I think it could be implemented so that if the policy became non-compliant again during the `delayAfterRunSeconds`, that event would be ignored and not trigger a re-run. It would only be re-run if it transitioned from compliant->noncompliant *after* `delayAfterRunSeconds`.

Like Matt mentioned, if the main use of PolicyAutomation is quick and used to create an alert or a ticket, then re-running like the current proposal is fine. But if the PolicyAutomation is meant to resolve the violation permanently, running a complex playbook each time, then running it on a loop could cause other issues.

I see potential advantages and disadvantages either way, so I just want to be sure it's discussed now.

**Review by JustinKuli — APPROVED** [review]
I can be ok with it as is. But I would appreciate another set of eyes on the potential re-run loop situation.

File: enhancements/sig-policy/59-ansible-everyevent-mode/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (mprahl) [reply enhancements/sig-policy/59-ansible-everyevent-mode/README.md:134]: > To me it sounds like you are queueing up to 1 action during the delay time

That's right. Let me try to reword this to make it clearer.

[the remark this answers is not in the document]
Comment (gparvin) [reply enhancements/sig-policy/59-ansible-everyevent-mode/README.md:134]: Thanks, yes.

**Review by JustinKuli — APPROVED** [review]

**Review by dhaiducek — APPROVED** [review]
